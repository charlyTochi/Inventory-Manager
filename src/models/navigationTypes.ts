import {NativeStackScreenProps} from '@react-navigation/native-stack';

export type ListItemStackParamList = {
  Splash: undefined;
  Login: undefined;
  ListItems: undefined;
  CreateItem: undefined;
  ViewItem: {id: string};
  EditItem: {id: string};
};


export type LoginScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'Login'
>;

export type ListItemsScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'ListItems'
>;
export type CreateItemScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'CreateItem'
>;
export type ViewItemScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'ViewItem'
>;
export type EditItemScreenProps = NativeStackScreenProps<
  ListItemStackParamList,
  'EditItem'
>;
