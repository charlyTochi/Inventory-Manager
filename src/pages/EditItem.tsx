import React, {useEffect, useState} from 'react';
import {
  ActivityIndicator,
  Alert,
  Dimensions,
  StyleSheet,
  View,
} from 'react-native';
import {IInventoryItem, Inventory} from '../services/inventory';
import {EditItemScreenProps} from '../models/navigationTypes';
import Form from '../components/Form';
import * as Yup from 'yup';

const inventory = new Inventory();
const {height} = Dimensions.get('window');

const EditItem: React.FC<EditItemScreenProps> = ({navigation, route}) => {
  const [item, setItem] = useState<IInventoryItem | null>(null);
  const [initialValues, setInitialValues] = useState({
    name: '',
    description: '',
    totalStock: '',
    price: '',
  });

  useEffect(() => {
    const fetchItem = async () => {
      const fetchedItem = await inventory.read(route.params.id);
      if (fetchedItem) {
        setItem(fetchedItem);
        setInitialValues({
          name: fetchedItem.name,
          description: fetchedItem.description,
          totalStock: fetchedItem.totalStock.toString(),
          price: fetchedItem.price.toString(),
        });
      }
    };

    fetchItem();
  }, [route.params.id]);

  if (!item) {
    return (
      <View style={styles.noDataDiv}>
        <ActivityIndicator />
      </View>
    );
  }

  const handleUpdateItem = async (values: any) => {
    try {
      await inventory.update(item.id, {
        name: values.name,
        totalStock: parseInt(values.totalStock, 10),
        price: parseFloat(values.price),
        description: values.description,
      });
      navigation.goBack();
    } catch (error: any) {
      Alert.alert('Error', error.message);
    }
  };

  const validationSchema = Yup.object().shape({
    name: Yup.string().required('Name is required'),
    description: Yup.string()
      .min(3, 'Name must be at least 3 characters long')
      .required('Description is required'),

    totalStock: Yup.number()
      .required('Total stock is required')
      .positive('Total stock must be positive'),
    price: Yup.number()
      .required('Price is required')
      .positive('Price must be positive'),
  });

  return (
    <View>
      {initialValues.name && (
        <Form
          initialValues={initialValues}
          onSubmit={handleUpdateItem}
          validationSchema={validationSchema}
          submitButtonText="Save"
        />
      )}
    </View>
  );
};

export default EditItem;

const styles = StyleSheet.create({
  noDataDiv: {
    position: 'relative',
    top: height / 2.5,
    bottom: height / 2,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
