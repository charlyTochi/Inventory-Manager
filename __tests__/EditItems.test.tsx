import React from 'react';
import {render} from '@testing-library/react-native';
import EditItem from '../src/pages/EditItem';
import {ListItemStackParamList} from '../src/models/navigationTypes';
import {RouteProp} from '@react-navigation/native';

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
}));

describe('EditItem screen', () => {
  const mockNavigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setOptions: jest.fn(),
    addListener: jest.fn(),
    removeListener: jest.fn(),
    reset: jest.fn(),
    pop: jest.fn(),
    popToTop: jest.fn(),
    dispatch: jest.fn(),
  };

  const mockRoute: RouteProp<ListItemStackParamList, 'EditItem'> = {
    key: 'EditItem',
    name: 'EditItem',
    params: {id: '1'},
  };

  it('should match the snapshot', () => {
    const {toJSON} = render(
      <EditItem navigation={mockNavigation as any} route={mockRoute} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
