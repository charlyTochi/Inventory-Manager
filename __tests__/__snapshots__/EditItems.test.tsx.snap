// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditItem screen should match the snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "bottom": 667,
      "justifyContent": "center",
      "position": "relative",
      "top": 533.6,
    }
  }
>
  <ActivityIndicator />
</View>
`;
